### PR TITLE
Design doc for Model average(renaming it to Parameter Average)

### PR DESCRIPTION
**Pull request by kavyasrinet**

### CODE REVIEW

File: doc/design/parameter_average.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# Averaging Parameter Updates in PaddlePaddle
+
+## Why Averaging
+In large scale machine learning, it could take us a large number of iterations over the training data to achieve optimal values of parameters of our model. The size of training data could be very large for large scale training hence it is desirable if we can obtain the optimal parameters by going through the data in as few passes as we can.
+Polyak and Juditsky (1992) showed that asymptotically the test performance of simple average of parameters obtained by Stochastic Gradient Descent (SGD) is as good as that of parameters which minimize the empirical cost.
```
Comment (pengli09) [inline doc/design/parameter_average.md]: It seems that this paragraph comes from the abstract of [this paper](https://arxiv.org/pdf/1107.2490.pdf) by Wei Xu. I think we need to add a citation here or rephrase this paragraph.

File: doc/design/parameter_average.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+Hence, overall we have have two copies of the parameters: one for the optimizer itself, and one for the ParameterAverageOptimizer. The former should be used in back propagation, while the latter should be used during testing and should be saved.
+
+During the testing/ saving the model phase, we perform the following steps:
+1. Perform the delayed operations.
+2. Save current values of the parameters to a temporary variable.
+3. Replace the values of the parameters with the averaged values.
+4. Perform testing and/or save the parameters.
+5. Restore the values of the parameters once done.
+
+### How to do implement Averaging of Parameter Updates in PaddlePaddle
```
Comment (pengli09) [inline doc/design/parameter_average.md]: Drop "do"
Drop "Updates"? -> see above

File: doc/design/parameter_average.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+We can add the ParameterAverageOptimizer op to the graph through Python API. Using this approach, we manually add this op to the graph and direct the output of the optimizer op to this op during training.
+
+	**Advantages**:
+    - Allows for greater flexibility to the users of Paddle. Using this approach, the users can plug different optimizers into ParameterAverageOptimizer by passing in the optimizer to the op.
```
Comment (pengli09) [inline doc/design/parameter_average.md]: s/Paddle/PaddlePaddle/g

File: doc/design/parameter_average.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+### Low-Level implementation
+
+In the new design, we propose to create a new operation for averaging parameter updates (ParameterAverageOptimizer). For now, we can add an op that takes in the following as input:
+- the optimizer
+- the window_size to keep the updates
+
+The ParameterAverageOptimizer op can be like any other operator with its own CPU/GPU implementation either using Eigen or separate CPU and GPU kernels. As the initial implementation, we can implement the kernel using Eigen following the abstraction pattern implemented for [Operators](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/rmsprop_op.h).
+
+The idea of building an op for averaging is in sync with the refactored Paddle philosophy of using operators to represent any computation unit. The way the op will be added to the computation graph will be decided by the [layer functions](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/python_api.md#layer-function) in Python API.
```
Comment (pengli09) [inline doc/design/parameter_average.md]: s/Paddle/PaddlePaddle/g

File: doc/design/parameter_average.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	**Disadvantages**:
+    - Implementation requires re-writing the averaging methodology in Python.  
+
+### Low-Level implementation
+
+In the new design, we propose to create a new operation for averaging parameter updates (ParameterAverageOptimizer). For now, we can add an op that takes in the following as input:
+- the optimizer
+- the window_size to keep the updates
+
+The ParameterAverageOptimizer op can be like any other operator with its own CPU/GPU implementation either using Eigen or separate CPU and GPU kernels. As the initial implementation, we can implement the kernel using Eigen following the abstraction pattern implemented for [Operators](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/rmsprop_op.h).
```
Comment (pengli09) [inline doc/design/parameter_average.md]: We should also support the case that trainer runs on GPU while ParameterAverageOptimizer runs on CPU in order to save GPU memory.

File: doc/design/parameter_average.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+### Python API implementation for ParameterAverageOptimizer
+
+Based on Polyak and Juditsky (1992), we can generalize the averaging of updates to any optimizer. The input to the op would be the following:
+- Any optimizer (RMSProp , AdaGrad etc.)
+- A window size. The op keeps accumulating updated parameter values over a window of N batches and takes an average. Move the averaged value to a buffer when window is full to avoid loss of precision.
+
+Using the ParameterAverageOptimizer op, any user can add the operation to their computation graphs. However, this will require a lot of lines of code and we should design Python APIs that support averaging. As per the PaddlePaddle [Python API design](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/python_api.md), the layer functions are responsible for creating operators, operator parameters and variables. Since parameter averaging will be an operator, it makes sense to create it in the layer functions.
```
Comment (pengli09) [inline doc/design/parameter_average.md]: Shall we still do the actual computation in C++? I think the computational cost will be very high if we do it in Python.

File: doc/design/parameter_average.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
+# Averaging Parameter Updates in PaddlePaddle
```
[the remark this answers is not in the document]
Comment (kavyasrinet) [reply doc/design/parameter_average.md]: Done.

File: doc/design/parameter_average.md (hunk 1 above)
Comment (kavyasrinet) [reply doc/design/parameter_average.md]: Done.

File: doc/design/parameter_average.md (hunk 2 above)
Comment (kavyasrinet) [reply doc/design/parameter_average.md]: Done.

File: doc/design/parameter_average.md (hunk 3 above)
Comment (kavyasrinet) [reply doc/design/parameter_average.md]: Done.

File: doc/design/parameter_average.md (hunk 5 above)
Comment (kavyasrinet) [reply doc/design/parameter_average.md]: Done.

File: doc/design/parameter_average.md (hunk 4 above)
Comment (kavyasrinet) [reply doc/design/parameter_average.md]: Done.

File: doc/design/parameter_average.md (hunk 6 above)
Comment (kavyasrinet) [reply doc/design/parameter_average.md]: Done.

**Pull request merged**